tiledb/sm/query/readers/aggregators/min_max_aggregator.cc

```diff
@@ -0,0 +1,413 @@
+/**
+ * @file min_max_aggregator.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class MinMaxAggregator.
+ */
+
+#include "tiledb/sm/query/readers/aggregators/min_max_aggregator.h"
+
+#include "tiledb/sm/array_schema/array_schema.h"
+#include "tiledb/sm/query/query_buffer.h"
+#include "tiledb/sm/query/readers/aggregators/aggregate_buffer.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class MinMaxAggregatorStatusException : public StatusException {
+ public:
+  explicit MinMaxAggregatorStatusException(const std::string& message)
+      : StatusException("MinMaxAggregator", message) {
+  }
+};
+
+template <typename T>
+MinMaxAggregator<T>::MinMaxAggregator(
+    const bool min,
+    const std::string field_name,
+    const bool var_sized,
+    const bool is_nullable,
+    const unsigned cell_val_num)
+    : is_nullable_(is_nullable)
+    , var_sized_(var_sized)
+    , cell_val_num_(cell_val_num)
+    , field_name_(field_name)
+    , value_(nullopt)
+    , validity_value_(is_nullable ? std::make_optional(0) : nullopt) {
+  if (var_sized && !std::is_same<T, std::string>::value) {
+    throw MinMaxAggregatorStatusException(
+        "Min/max aggregates must not be requested for var sized non-string "
+        "attributes.");
+  }
+
+  if (cell_val_num != 1 && !std::is_same<T, std::string>::value) {
+    throw MinMaxAggregatorStatusException(
+        "Min/max aggregates must not be requested for attributes with more "
+        "than one value.");
+  }
+
+  if (min) {
+    op_ = std::less<VALUE_T>();
```

Comment:
For discussion, but strongly lean toward calculating both values at once. Most consumers are going to want both, and it is much cheaper to calculate them together rather than separately.